### ISSUE TITLE
Task and Future

### ISSUE THREAD

**Issue by rlavolee** [spebbe/dartz#24]
Hi spebbe,

In the first place, I want to thank you to bring functional programing to Dart.
However I have a question related to `Task` and `Future`. If you got time to take a look at my problem I will be thankfull.

I am running this code with no success and I don't have any clue left.

```dart
Task<Option<Product>> getProduct(int id) =>
      Task(() => this.httpClient.get(someUrl))
        .map((resp) => Some(ProductFromResp(resp)));
```

As I understand  in this case Task takes a `Function0 Future<Response>` as a parameter and map the `http response` to produce an `Option<Product>`.  However when I explicitly run `getProduct(id).run()` the task it doesn't seem to be fired. 

Am I missing something ?

Regards

**Comment by spebbe** [comment]
Hi!

That should work...

My first guess is that `this.httpClient.get(someUrl)` throws an exception that is being ignored after it is rethrown by `Task#run`. Task just delays execution of Future and puts a monadic interface on top -- it doesn't encapsulate exceptions any more than Future already does.

Does that sound like a possible explanation? If not, try replacing `this.httpClient.get(someUrl)` with something like `Future(() => mockedResponse)` and see if you can figure out the issue from there?

**Comment by rlavolee** [comment]
Thank you very much for your reply ! 

Your suggestion is a good idea. I tried it. 

```
Task(() => Future(() => 42))
        .map((resp) => Some(ProductFromResp(resp)));
```

However the task didn't seem to be fired. Even good old fashion `print` didn't print anything. So I tried a simple `Task.delay`.

```
Task.delay(() => Some(ProductFromResp(42)));
```

And everything was fine. 
So it seems only behaved oddly with `Future` in `Task` constructor.
What do you think ? (BTW I'm using dartz for a flutter app.)

**Comment by spebbe** [comment]
Really odd... I'll try to reproduce and get back to you.

**Comment by rlavolee** [comment]
I quickly wrote this test and it passed...

```
import 'dart:async';

import 'package:dartz/dartz.dart';
import 'package:test/test.dart';

void main() {

  test("Task should return 42", () async {
    expect(Task.delay(() => 42).run(), completion(equals(42)));
  });

  test("Task should return 42 from Future as contructor parameter", () async {
    final Task t = Task(() => Future(() => 42));
    expect(t.run(), completion(equals(42)));
  });
} 
```

**Comment by rlavolee** [comment]
Hi spebbe, I think I finally located the problem. 
For some reason, in Flutter the `map` function is not triggered. I replaced `map` with `bind` and it worked. I guess this is because `Task` is run directly with `bind`.

¯\\_(ツ)_/¯

**Comment by spebbe** [comment]
@rlavolee I tried this on flutter 1.7.8+hotfix.4 and dartz 0.8.6:
```
final t = Task(() => Future(() => 42)).map(some);
final v = await t.run();
print(v);
```
For me, it prints the expected `Some(42)`... no idea why it misbehaves for you?

**Comment by rlavolee** [comment]
> @rlavolee I tried this on flutter 1.7.8+hotfix.4 and dartz 0.8.6:
> 
> ```
> final t = Task(() => Future(() => 42)).map(some);
> final v = await t.run();
> print(v);
> ```
> 
> For me, it prints the expected `Some(42)`... no idea why it misbehaves for you?

This case also works for me and the following

```dart
Task<Option<int>> getInt(int i) => Task.delay(() => i);

final t = getInt(42).map(some);
final v = await t.run();
print(v);
```

However not this one. 🤔 

```dart
Task<Option<int>> getInt(int i) => Task.delay(() => i).map(some);

final t = getInt(42);
final v = await t.run();
print(v);
```

**Comment by rlavolee** [comment]
Hi spebbe,

I removed the type annotation and it finally works.
Thanks for your help.

**Comment by spebbe** [comment]
OK, here's what I think happens... The `Task#map` operation on the 0.8.x series (=master) isn't Dart 2 clean yet, causing runtime type checks to fail in the second of your examples. In the first example, the type checks succeed because type information is lost -- the type of `t` is inferred as `Task<dynamic>` 😩

This is already fixed on the infamous `0_9_0_wip_2` branch, but I'll backport the fix to master and release a new stable version in the mean time.

**Comment by spebbe** [comment]
Reopening because this really is a bug on Dart 2. Thanks for the help in narrowing it down!

**Comment by rlavolee** [comment]
Hi @spebbe ! Can I help in a way ?

**Comment by spebbe** [comment]
@rlavolee, I've backported the fixes to `Task` from the 0.9.x branch to master and released a new version 0.8.7, which should resolve the problems you were seeing as well as give you better type inference.
I'm closing this issue for now, but please reopen if you are still having problems!
